### PR TITLE
Fixes floating point error in ELM for BGC compsets

### LINKED ISSUE #2832: floating invalid in subgridAveMod.F90 using DEBUG=TRUE and force 1 thread on cori
Using master and the following test:

`ERS_D_PMx1.f19_g16.I1850CNECACNTBC.cori-knl_intel.clm-eca`

(which is DEBUG=TRUE and force 1 thread), I get the following error:

```
  5: forrtl: error (65): floating invalid
  5: Image              PC                Routine            Line        Source
  5: e3sm.exe           0000000006386B7E  Unknown               Unknown  Unknown
  5: e3sm.exe           0000000005C293D0  Unknown               Unknown  Unknown
  5: e3sm.exe           0000000000E31C9B  subgridavemod_mp_         238  subgridAveMod.F90
  5: e3sm.exe           000000000391A922  vegetationdatatyp        3592  VegetationDataType.F90
  5: e3sm.exe           00000000008C6F71  clm_driver_mp_clm         355  clm_driver.F90
  5: e3sm.exe           000000000084F1AA  lnd_comp_mct_mp_l         509  lnd_comp_mct.F90
  5: e3sm.exe           0000000000461A94  component_mod_mp_         724  component_mod.F90
  5: e3sm.exe           000000000042CBBC  cime_comp_mod_mp_        2447  cime_comp_mod.F90
  5: e3sm.exe           000000000044A3B8  MAIN__                    133  cime_driver.F90
```

I see two source files with same name:  
```
./components/clm/src/main/subgridAveMod.F90
./components/clm/src/external_models/sbetr/src/stub_clm/subgridAveMod.F90
```

I get the same error when NOT forcing 1 thread, but this rules out a threading issue.
I get the same error with the following tests:

```
ERS_D.f19_f19.I20TRCLM45CN.cori-knl_intel
ERS_D.f19_g16.I1850CNECACNTBC.cori-knl_intel.clm-eca
ERS_D.f19_g16.I1850GSWCNPECACNTBC.cori-knl_intel.clm-eca_f19_g16_I1850GSWCNPECACNTBC

ERS_D_PMx1.f19_f19.I20TRCLM45CN.cori-knl_intel
ERS_D_PMx1.f19_g16.I1850CNECACNTBC.cori-knl_intel.clm-eca
ERS_D_PMx1.f19_g16.I1850GSWCNPECACNTBC.cori-knl_intel.clm-eca_f19_g16_I1850GSWCNPECACNTBC
```

Example case dir:
`/global/cscratch1/sd/ndk/acme_scratch/cori-knl/m99-apr1/ERS_D_PMx1.f19_g16.I1850CNECACNTBC.cori-knl_intel.clm-eca.20190401_132358_h3vcag`

### PR DESCRIPTION
**Pull request by jinyuntang**
Fixed a number of initialization issues related to the logical switch of crop model and dynamic land use.  

Fixes #2832
[BFB]

### CODE REVIEW

**Comment by jinyuntang** [comment]
@ndkeen and @bishtgautam, I have run the failed tests mentioned in #2832, they all passed after this fix.

File: components/clm/src/data_types/ColumnDataType.F90 [hunk 1]
```diff
@@ -7418,7 +7418,7 @@ subroutine col_nf_init(this, begc, endc)
     allocate(this%col_plant_ndemand_vr            (begc:endc,1:nlevdecomp))       ; this%col_plant_ndemand_vr          (:,:) = nan
     allocate(this%col_plant_nh4demand_vr          (begc:endc,1:nlevdecomp))       ; this%col_plant_nh4demand_vr        (:,:) = nan
     allocate(this%col_plant_no3demand_vr          (begc:endc,1:nlevdecomp))       ; this%col_plant_no3demand_vr        (:,:) = nan
-    allocate(this%col_plant_pdemand_vr            (begc:endc,1:nlevdecomp))       ; this%col_plant_pdemand_vr          (:,:) = nan
+    allocate(this%col_plant_pdemand_vr            (begc:endc,1:nlevdecomp))       ; this%col_plant_pdemand_vr          (:,:) = 0._r8
```
Comment (bishtgautam) [inline components/clm/src/data_types/ColumnDataType.F90]: Why set the col_plant_pdemand_vr to 0 instead of nan?

**Comment by bishtgautam** [comment]
Do you know why a similar change is not needed for the N variable (i.e. `col_plant_ndemand_vr`)?

**Comment by jinyuntang** [comment]
The variable for N is used at the same time step as it is reset to zero.

**Comment by bpbond** [comment]
Given recent re-architecting, is useful to request a review from Peter as well?

**Comment by bishtgautam** [comment]
I have now added @thorntonpe as a reviewer.

**Comment by jinyuntang** [comment]
@thorntonpe, do you have any comments? Thanks

**Comment by thorntonpe** [comment]
@jinyuntang I have looked this over, and have some notes/suggestions. First, the use of the variable col_plant_pdemand_vr from the column_nitrogen_flux data structure (in ColumnDataType.F90) is restricted to the ECA case. When I made the architecture migration, I moved all the variables that were in the original nitrogenflux data structure into their respective subgrid-level data structures, and I failed to notice that this variable, which is a phosphorus flux variable, existed within the nitrogenflux data structure. So I moved it into the column nitrogen flux structure instead of into the column phosphorus flux structure. It looks to me like the usage of this variable within the ECA blocks of AllocationMod.F90 is fine, but the location of the variable is confusing. So, since you are working on this, my first suggestion is to move this variable into the column_phosphorus_flux data type. Then, instead of setting the value to 0.0 within the allocate block, I would use the nan value as for the other allocates, but then add a new line for col_plant_pdemand_vr with the correct indexing within the col_pf_setvalues() subroutine, also in ColumnDataType.F90.

**Comment by thorntonpe** [comment]
@jinyuntang... also...
It looks like you are deleting the initialization for the variable plant_demand_vr, which is already a part of the column_phosphorus_flux data type, in the col_pf_setvalues() subroutine, at line 9764. This should not be done, since that variable is used in the CTC code (but not in the ECA code, I think), and needs to be initialized during the call to this setvalues routine. To clarify: ECA is using the variable col_plant_pdemand_vr, currently define in the column_nitrogen_flux structure, while CTC is using the variable plant_pdemand_vr, currently defined in the column_phosphorus_flux structure. When you move the ECA variable into the correct structure, you should keep the name col_plant_pdemand_vr to avoid confusion between the two variables. Then you should add the appropriate initialization for that variable in the col_pf_setvalues() routine, without making any changes to the handling of plant_pdemand_vr. Once you do that, you'll need to go into AllocationMod.F90, and find the couple of places where the col_plant_pdemand_vr variable is used in an associate statement, and replace the col_nf% prefix with col_pf%.

**Comment by jinyuntang** [comment]
@thorntonpe, thank for your careful checking this. I will fix the issues you identified above.

**Comment by thorntonpe** [comment]
@jinyuntang, one further note. I'm not sure about the modifications you are suggesting in dynSubgridAdjustmentsMod.F90. The addition of iulog on line 30 and the addition of the extra index variable c on line 854 seem unnecessary. The changes on lines 864 and 1290, which set the adjustment state variables to zero before each pass through the subroutine seem likely to change answers if the subroutine is called more than once. I have not traced this, but I think you should explain why you are adding this change. It may not have an impact until there is dynamic landunit area.

**Comment by thorntonpe** [comment]
@jinyuntang one last question/comment - have you run the land developer test suite against master baselines for this PR? I would have thought the change in initialization of plant_pdemand_vr variable would cause a fail against baselines.

**Comment by jinyuntang** [comment]
@thorntonpe, the addition of iulog and c were leftover I used to tracking down the error, and was thinking they will be used again for further "development and testing" purpose, but I am OK to remove them. For the changes in changes on lines 864 and 1290, they are used to ensure a same initialization as done for col_cs%dyn_cbal_adjustments. I followed Noel's suggestion only ran the failed cases. But I can run the full test suite after revising the point you raised above.

**Comment by jinyuntang** [comment]
@thorntonpe, I have updated the branch according to your suggestions and ran the test suite e3sm_developer. All tests looked good. Please take another look. Thanks.

**Comment by thorntonpe** [comment]
@jinyuntang thanks, the changes look good to me. Did your e3sm_developer test include a baseline comparison against the branch point on master?

**Comment by jinyuntang** [comment]
@thorntonpe, I did check one for which I expect some changes may occur. But jus in case, do you have an instruction for a more comprehensive test (include more wall to wall comparisons as you implied) other than  e3sm_developer?  Thanks

**Comment by jinyuntang** [comment]
@thorntonpe, @bishtgautam and @ndkeen, I ran the e3sm_land_developer test suit compared between my branch and its baseline, and found only the following complains.
    FAIL ERS.f09_g16.I1850CLM45CN.cori-knl_intel TPUTCOMP Error: Computation time increase > 10 pct from baseline
    FAIL ERS.f19_g16.I20TRGSWCNPRDCTCBC.cori-knl_intel.clm-ctc_f19_g16_I20TRGSWCNPRDCTCBC TPUTCOMP Error: Computation time increase > 10 pct from baseline
    FAIL ERS.ne11_oQU240.I20TRCLM45.cori-knl_intel TPUTCOMP Error: Computation time increase > 10 pct from baseline

Should this satisfy the BFB criteria? 

Thanks

**Comment by thorntonpe** [comment]
@jinyuntang yes, that satisfies the BFB criteria - thanks!

**Pull request merged**